### ISSUE TITLE
Enable GPU sharing on test cluster

### ISSUE THREAD

**Issue by larsks** [nerc-project/operations#468]
Currently, only a single Pod can use a GPU. In many cases this will be inefficient (because a single Pod may not fully utilize the GPU). There are a couple of different options for sharing a GPU among multiple workloads:

MIG (multi instance gpu):

- [RedHat MIG article](https://www.redhat.com/en/blog/multi-instance-gpu-support-with-the-gpu-operator-v1.7.0)
- [NVIDIA MIG documentation](https://docs.nvidia.com/datacenter/cloud-native/openshift/latest/mig-ocp.html)

Time slicing:

- [Time slicing NVIDIA GPUs in OpenShift](https://docs.nvidia.com/datacenter/cloud-native/openshift/23.9.1/time-slicing-gpus-in-openshift.html)

We should experiment with these configurations on the test cluster, particularly in the context of OpenShift AI (nèe RHODS).

**Comment by hpdempsey** [comment]
Ahmed is also interested in methods for sharing GPUs.   He's not a member of this repo, but I've passed him the pointer to this testing, in case he wants to also do some work in the test cluster with them.

**Comment by dystewart** [comment]
We should note there are 3 different MIG advertisement strategies (single, multi, none) configured [here](https://github.com/OCP-on-NERC/nerc-ocp-config/blob/bb2ce449f504c8d88f1643053de6aed88ef9c676/nvidia-gpu-operator/base/gpu-cluster-policy.yaml#L58), should we want to change this in the future we would need to create a new GPU Operator ClusterPolicy as "updates to the ClusterPolicy after it has been deployed are currently silently ignored."

**Comment by dystewart** [comment]
Our current MIG strategy is single, which means: `In this strategy, all the GPUs in a single node are configured in a homogenous manner (same number of compute units, same memory size).`

So if we chose this geometry: 1g.5gb = 7 distinct nvidia.com/gpu instances 

This seems like the natural approach for testing at least. 

The alternative is mixed the difference being: `There is no constraint on the geometry`

**Comment by hpdempsey** [comment]
I think we will need to test both modes, but staring with multi makes sense.  Four GPUs are probably more expensive than all PIs can afford, so that won't be flexible enough as our only policy.  Is the policy variable by node, or set for all GPUs in the cluster?

**Comment by dystewart** [comment]
@hpdempsey @schwesig I can now clarify a bit of what I stated above and introduce a bit of vocab for clarity.

I think you're right that for our purposes mixed will make the most sense. This allows us the ability "slice" the gpu nodes however we like. 
Slices: `are essentially virtual gpus that appear as individual gpu nodes, with predetemined geometry.`
Geometry: `the resource package (compute & memory) we assign to a slice, for instance: 1g.5gb`

**Comment by dystewart** [comment]
Next important thing, given that:

> We should note there are 3 different MIG advertisement strategies (single, multi, none) configured [here](https://github.com/OCP-on-NERC/nerc-ocp-config/blob/bb2ce449f504c8d88f1643053de6aed88ef9c676/nvidia-gpu-operator/base/gpu-cluster-policy.yaml#L58), should we want to change this in the future we would need to create a new GPU Operator ClusterPolicy as "updates to the ClusterPolicy after it has been deployed are currently silently ignored."

let's re-create the gpu cluster policy to reflect the "mixed" mig mode (*working on PR for this*)

**Comment by dystewart** [comment]
We've had success slicing gpu with single MIG strategy, with geometry: 1g.5gb. We had some confusion over how many MIG devices this would create (in the docs a node refers to a single gpu not the NVIDIA device as a whole). So instead of only 7 MIGs we get 4 * 7 = 28 MIGs

**Comment by dystewart** [comment]
Here is a what all the devices look like: 
```
Thu Mar 14 20:16:24 2024       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.54.14              Driver Version: 550.54.14      CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA A100-SXM4-40GB          On  |   00000000:31:00.0 Off |                   On |
| N/A   25C    P0             69W /  400W |                  N/A   |     N/A      Default |
|                                         |                        |              Enabled |
+-----------------------------------------+------------------------+----------------------+
|   1  NVIDIA A100-SXM4-40GB          On  |   00000000:4B:00.0 Off |                   On |
| N/A   25C    P0             72W /  400W |                  N/A   |     N/A      Default |
|                                         |                        |              Enabled |
+-----------------------------------------+------------------------+----------------------+
|   2  NVIDIA A100-SXM4-40GB          On  |   00000000:CA:00.0 Off |                   On |
| N/A   25C    P0             68W /  400W |                  N/A   |     N/A      Default |
|                                         |                        |              Enabled |
+-----------------------------------------+------------------------+----------------------+
|   3  NVIDIA A100-SXM4-40GB          On  |   00000000:E3:00.0 Off |                   On |
| N/A   25C    P0             70W /  400W |                  N/A   |     N/A      Default |
|                                         |                        |              Enabled |
+-----------------------------------------+------------------------+----------------------+

+-----------------------------------------------------------------------------------------+
| MIG devices:                                                                            |
+------------------+----------------------------------+-----------+-----------------------+
| GPU  GI  CI  MIG |                     Memory-Usage |        Vol|      Shared           |
|      ID  ID  Dev |                       BAR1-Usage | SM     Unc| CE ENC DEC OFA JPG    |
|                  |                                  |        ECC|                       |
|==================+==================================+===========+=======================|
|  0    7   0   0  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  0    8   0   1  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  0    9   0   2  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  0   10   0   3  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  0   11   0   4  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  0   12   0   5  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  0   13   0   6  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  1    7   0   0  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  1    8   0   1  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  1    9   0   2  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  1   10   0   3  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  1   11   0   4  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  1   12   0   5  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  1   13   0   6  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  2    7   0   0  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  2    8   0   1  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  2    9   0   2  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  2   10   0   3  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  2   11   0   4  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  2   12   0   5  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  2   13   0   6  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  3    7   0   0  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  3    8   0   1  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  3    9   0   2  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  3   10   0   3  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  3   11   0   4  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  3   12   0   5  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  3   13   0   6  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
                                                                                         
+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|  No running processes found                                                             |
+-----------------------------------------------------------------------------------------+
```

**Comment by dystewart** [comment]
@DanNiESh ran the same test script that generated the above output in a notebook with a single MIG allocated: 
```
(app-root) (app-root) nvidia-smi
Thu Mar 14 20:56:37 2024       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.54.14              Driver Version: 550.54.14      CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA A100-SXM4-40GB          On  |   00000000:31:00.0 Off |                   On |
| N/A   25C    P0             69W /  400W |                  N/A   |     N/A      Default |
|                                         |                        |              Enabled |
+-----------------------------------------+------------------------+----------------------+

+-----------------------------------------------------------------------------------------+
| MIG devices:                                                                            |
+------------------+----------------------------------+-----------+-----------------------+
| GPU  GI  CI  MIG |                     Memory-Usage |        Vol|      Shared           |
|      ID  ID  Dev |                       BAR1-Usage | SM     Unc| CE ENC DEC OFA JPG    |
|                  |                                  |        ECC|                       |
|==================+==================================+===========+=======================|
|  0    7   0   0  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
                                                                                         
+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|  No running processes found                                                             |
+-----------------------------------------------------------------------------------------+
```

**Comment by dystewart** [comment]
We also tested a a load on the gpu suggested in the NVIDIA OCP docs. This shows how to allocate a gpu node manually outside of rhods:

```
apiVersion: v1
kind: Pod
metadata:
  name: cuda-vectoradd
spec:
  restartPolicy: OnFailure
  containers:
  - name: cuda-vectoradd
    image: "nvidia/samples:vectoradd-cuda11.2.1"
    resources:
      limits:
        nvidia.com/gpu: 1
  nodeSelector:
    nvidia.com/gpu.product: NVIDIA-A100-SXM4-40GB-MIG-1g.5gb
```

Which yielded successful run:
```
[Vector addition of 50000 elements]
Copy input data from the host memory to the CUDA device
CUDA kernel launch with 196 blocks of 256 threads
Copy output data from the CUDA device to the host memory
Test PASSED
Done
```

**Comment by dystewart** [comment]
So now that we are happy with the single strategy I'll make a PR adding a few resources as well as updating the MIG strategy to mixed as discussed above: 


> Next important thing, given that:
> 
> > We should note there are 3 different MIG advertisement strategies (single, multi, none) configured [here](https://github.com/OCP-on-NERC/nerc-ocp-config/blob/bb2ce449f504c8d88f1643053de6aed88ef9c676/nvidia-gpu-operator/base/gpu-cluster-policy.yaml#L58), should we want to change this in the future we would need to create a new GPU Operator ClusterPolicy as "updates to the ClusterPolicy after it has been deployed are currently silently ignored."
> 
> let's re-create the gpu cluster policy to reflect the "mixed" mig mode (_working on PR for this_)

**Comment by syockel** [comment]
This is great to see!!!

**Comment by joachimweyl** [comment]
A question worth keeping in mind is how we will differentiate what each pod is using. Is there currently an easy way to know that a pod is using a MIG slice instead of the entire GPU? PCI Label or something similar?

**Comment by DanNiESh** [comment]
We can find the GPU size in pods yaml file, this is the pod running with a MIG slice:
```
  containers:
    - resources:
        limits:
          cpu: '1'
          memory: 2Gi
          nvidia.com/gpu: '1'
        requests:
          cpu: '1'
          memory: 2Gi
          nvidia.com/gpu: '1'
```

After MIG is enabled, GPU unit is 1/7 GPU, so one GPU means one slice of GPU. If you need a whole GPU I think you need to claim 7 GPU (slices).

**Comment by dystewart** [comment]
@joachimweyl I think this might also help to answer your question: 


> We also tested a a load on the gpu suggested in the NVIDIA OCP docs. This shows how to allocate a gpu node manually outside of rhods with `nodeSelector`:
> 
> ```
> apiVersion: v1
> kind: Pod
> metadata:
>   name: cuda-vectoradd
> spec:
>   restartPolicy: OnFailure
>   containers:
>   - name: cuda-vectoradd
>     image: "nvidia/samples:vectoradd-cuda11.2.1"
>     resources:
>       limits:
>         nvidia.com/gpu: 1
>   nodeSelector:
>     nvidia.com/gpu.product: NVIDIA-A100-SXM4-40GB-MIG-1g.5gb

Each GPU is labeled with it's device and geometry (see the nodeSelecter field above)

See also under resources.limits.nvidia.com/gpu:1
This tells us that we're allocating 1 gpu from the node labeled: NVIDIA-A100-SXM4-40GB-MIG-1g.5gb
In this case that allocates a MIG device that is a 1/7 slice of the total GPU

**Comment by joachimweyl** [comment]
Once a GPU has MIG enabled can it still also be used as a whole single GPU or a whole 2, 4 GPUs if that is what the user wants?

**Comment by joachimweyl** [comment]
Is this issue for testing as well as enabling or do we need a new issue for testing?

**Comment by DanNiESh** [comment]
> also

From my understanding, once the MIG is enabled, each GPU slice is considered as an individual GPU. If a user need a whole GPU, we need to disable MIG. There is no way to serve MIG mode and full resource mode at the same time.

**Comment by DanNiESh** [comment]
> Is this issue for testing as well as enabling or do we need a new issue for testing? We have [this](https://github.com/CCI-MOC/ops-issues/issues/1039) from a while back if we need a separate issue if not we can close that one with this one as the "duplicate".

I think these two issues are duplicate.

**Comment by dystewart** [comment]
> Once a GPU has MIG enabled can it still also be used as a whole single GPU or a whole 2, 4 GPUs if that is what the user wants?

Yes it can! with the mixed strategy you can slice the individual gpus however you'd like, not to mention they can be resliced very easily if the user wants a different geometry. We can control how things are sliced all the way down to the individual gpu device in the nvidia nodes, see [nvidia custom config sample](https://github.com/NVIDIA/mig-parted/blob/51b3d64749384857783dc8dca6b766851bf1ed17/examples/config.yaml#L50C1-L70C21)

**Comment by dystewart** [comment]
wrt the change made in: [https://github.com/OCP-on-NERC/nerc-ocp-config/pull/408](https://github.com/OCP-on-NERC/nerc-ocp-config/pull/408) we have tested the mixed mig mode with the following geometry: 
```
all-balanced-test:
        - devices: all
          mig-enabled: true
          mig-devices:
            "1g.5gb": 2
            "2g.10gb": 1
            "3g.20gb": 1
```

This results in 16 migs. Sample test run of command-nvidia-smi confirms:
```
Fri Mar 22 19:30:40 2024       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.54.14              Driver Version: 550.54.14      CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA A100-SXM4-40GB          On  |   00000000:31:00.0 Off |                   On |
| N/A   25C    P0             69W /  400W |                  N/A   |     N/A      Default |
|                                         |                        |              Enabled |
+-----------------------------------------+------------------------+----------------------+
|   1  NVIDIA A100-SXM4-40GB          On  |   00000000:4B:00.0 Off |                   On |
| N/A   25C    P0             72W /  400W |                  N/A   |     N/A      Default |
|                                         |                        |              Enabled |
+-----------------------------------------+------------------------+----------------------+
|   2  NVIDIA A100-SXM4-40GB          On  |   00000000:CA:00.0 Off |                   On |
| N/A   25C    P0             68W /  400W |                  N/A   |     N/A      Default |
|                                         |                        |              Enabled |
+-----------------------------------------+------------------------+----------------------+
|   3  NVIDIA A100-SXM4-40GB          On  |   00000000:E3:00.0 Off |                   On |
| N/A   25C    P0             70W /  400W |                  N/A   |     N/A      Default |
|                                         |                        |              Enabled |
+-----------------------------------------+------------------------+----------------------+

+-----------------------------------------------------------------------------------------+
| MIG devices:                                                                            |
+------------------+----------------------------------+-----------+-----------------------+
| GPU  GI  CI  MIG |                     Memory-Usage |        Vol|      Shared           |
|      ID  ID  Dev |                       BAR1-Usage | SM     Unc| CE ENC DEC OFA JPG    |
|                  |                                  |        ECC|                       |
|==================+==================================+===========+=======================|
|  0    2   0   0  |              37MiB / 19968MiB    | 42      0 |  3   0    2    0    0 |
|                  |                 0MiB / 32767MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  0    3   0   1  |              25MiB /  9856MiB    | 28      0 |  2   0    1    0    0 |
|                  |                 0MiB / 16383MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  0    9   0   2  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  0   10   0   3  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  1    2   0   0  |              37MiB / 19968MiB    | 42      0 |  3   0    2    0    0 |
|                  |                 0MiB / 32767MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  1    3   0   1  |              25MiB /  9856MiB    | 28      0 |  2   0    1    0    0 |
|                  |                 0MiB / 16383MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  1    9   0   2  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  1   10   0   3  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  2    2   0   0  |              37MiB / 19968MiB    | 42      0 |  3   0    2    0    0 |
|                  |                 0MiB / 32767MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  2    3   0   1  |              25MiB /  9856MiB    | 28      0 |  2   0    1    0    0 |
|                  |                 0MiB / 16383MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  2    9   0   2  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  2   10   0   3  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  3    2   0   0  |              37MiB / 19968MiB    | 42      0 |  3   0    2    0    0 |
|                  |                 0MiB / 32767MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  3    3   0   1  |              25MiB /  9856MiB    | 28      0 |  2   0    1    0    0 |
|                  |                 0MiB / 16383MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  3    9   0   2  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
|  3   10   0   3  |              12MiB /  4864MiB    | 14      0 |  1   0    0    0    0 |
|                  |                 0MiB /  8191MiB  |           |                       |
+------------------+----------------------------------+-----------+-----------------------+
                                                                                         
+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|  No running processes found                                                             |
+-----------------------------------------------------------------------------------------+
```

**Comment by dystewart** [comment]
A test gpu workload using:
```
apiVersion: v1
kind: Pod
metadata:
  name: cuda-vectoradd
spec:
  restartPolicy: OnFailure
  containers:
  - name: cuda-vectoradd
    image: "nvidia/samples:vectoradd-cuda11.2.1"
    resources:
      limits:
        nvidia.com/mig-2g.10gb: 1
      requests:
        nvidia.com/mig-2g.10gb: 1
```

Yields:
```
[Vector addition of 50000 elements]
Copy input data from the host memory to the CUDA device
CUDA kernel launch with 196 blocks of 256 threads
Copy output data from the CUDA device to the host memory
Test PASSED
Done
```

**Comment by dystewart** [comment]
So we are pretty happy with the functionality of the gpu sharing in the test cluster via mig partitions. Integrating migs to be useful with rhods however is a bit more complex because with the 16 migs we created above rhods only sees this: 

![image](https://github.com/nerc-project/operations/assets/84931428/8fa92710-8b58-4ac7-8048-2a466133e1ae)

Which is fine when all the migs are the same size (and we only have a single nvidia gpu in cluster), but with mixed strategy we cant differentiate what we're requesting out of the box. 

To differentiate migs in rhods we need to introduce acceleratorProfiles: [https://access.redhat.com/documentation/en-us/red_hat_openshift_data_science_self-managed/2.4/html/working_on_data_science_projects/working-with-accelerators_accelerators](https://access.redhat.com/documentation/en-us/red_hat_openshift_data_science_self-managed/2.4/html/working_on_data_science_projects/working-with-accelerators_accelerators)

In short it will allow us to go from showing all gpus as the same in rhods: 
```
spec:
  displayName: NVIDIA GPU
  enabled: true
  identifier: nvidia.com/gpu
  tolerations:
    - effect: NoSchedule
      key: nvidia.com/gpu
      operator: Exists
```

To distinguishing based on the mig node label:
```
spec:
  displayName: NVIDIA GPU
  enabled: true
  identifier: nvidia.com/mig-1g.10gb
  tolerations:
    - effect: NoSchedule
      key: nvidia.com/mig-1g.10gb
      operator: Exists
```

So next we will need to: 

1. implement and test the accelerators 
2. test time slicing on migs: [https://docs.nvidia.com/datacenter/cloud-native/openshift/23.9.1/time-slicing-gpus-in-openshift.html](https://docs.nvidia.com/datacenter/cloud-native/openshift/23.9.1/time-slicing-gpus-in-openshift.html)
